File: frontend/src/lib/types/compare.ts
export const CHANGE_VERB = {
	APPEARED: 'appeared',
	CHANGED: 'changed',
	DISAPPEARED: 'disappeared',
	UNCONFIRMED: 'unconfirmed',
	UNCHANGED: 'unchanged'
} as const;

export type ChangeVerb = (typeof CHANGE_VERB)[keyof typeof CHANGE_VERB];

export const VERB_ORDER: ChangeVerb[] = [
	CHANGE_VERB.APPEARED,
	CHANGE_VERB.CHANGED,
	CHANGE_VERB.DISAPPEARED,
	CHANGE_VERB.UNCONFIRMED,
	CHANGE_VERB.UNCHANGED
];

export const LISTED_VERBS: ChangeVerb[] = VERB_ORDER.slice(0, 4);

export const COMPARABILITY = {
	LIKE_FOR_LIKE: 'like_for_like',
	SETTINGS_DIFFER: 'settings_differ',
	QUALITY_DIFFERS: 'quality_differs',
	NOT_COVERED: 'not_covered'
} as const;

export type Comparability = (typeof COMPARABILITY)[keyof typeof COMPARABILITY];

export interface RunSide {
	scan_id: string;
	engine_name: string;
	context_name: string | null;
	intensity: string;
	scope: string;
	status: string;
	started_at: string | null;
	completed_at: string | null;
	duration_seconds: number | null;
	stages_ran: number;
	stages_planned: number;
	counts: Record<string, number>;
}

export interface StageDiff {
	name: string;
	title: string;
	baseline: string | null;
	current: string | null;
}

export interface SettingDiff {
	stage: string;
	title: string;
	field: string;
	label: string;
	before: string | null;
	after: string | null;
}

export interface RunDifference {
	key: string;
	label: string;
	baseline: string | null;
	current: string | null;
	material: boolean;
}

export interface CoverageLine {
	label: string;
	baseline: string | null;
	current: string | null;
}

export interface DimensionVerdict {
	dimension: string;
	comparability: Comparability;
	covered_baseline: boolean;
	covered_current: boolean;
	compared: boolean;
	confirmed: boolean;
	note: string;
	stages: string[];
	settings: SettingDiff[];
	coverage: CoverageLine[];
}

export interface DimensionDelta {
	dimension: string;
	label: string;
	noun: string;
	noun_plural: string;
	verdict: DimensionVerdict;
	total_baseline: number;
	total_current: number;
	appeared: number;
	changed: number;
	disappeared: number;
	unconfirmed: number;
	unchanged: number;
	intel_moved: number;
}

export interface ChangeFieldDelta {
	field: string;
	label: string;
	before: string | null;
	after: string | null;
	tone: 'up' | 'down' | 'neutral';
}

export interface ScreenshotPair {
	baseline: string | null;
	current: string | null;
}

export interface ChangeRow {
	key: string;
	dimension: string;
	verb: ChangeVerb;
	signal: string;
	rank: number;
	title: string;
	subtitle: string;
	scan_id: string;
	fields: ChangeFieldDelta[];
	severity: string | null;
	status: number | null;
	port: number | null;
	is_kev: boolean;
	sensitive: boolean;
	screenshots: ScreenshotPair | null;
}

export interface ChangeRows {
	dimension: string;
	items: ChangeRow[];
	total: number;
	page: number;
	size: number;
}

export interface ComparableRun {
	scan_id: string;
	engine_name: string;
	status: string;
	scope: string;
	started_at: string | null;
	duration_seconds: number | null;
	counts: Record<string, number>;
	dimensions: string[];
	comparable: boolean;
	reason: string;
}

export interface ScanComparison {
	target_id: string;
	target_value: string;
	target_type: string;
	baseline: RunSide;
	current: RunSide;
	dimensions: DimensionDelta[];
	stage_diff: StageDiff[];
	setting_diff: SettingDiff[];
	settings_identical: number;
	run_diff: RunDifference[];
	runs_between: number;
	comparability: Comparability;
	headline: string;
	summary: string;
	changes_total: number;
	live: boolean;
	suggestion: ComparableRun | null;
	generated_at: string;
}

export function listedCount(delta: DimensionDelta): number {
	return delta.appeared + delta.changed + delta.disappeared + delta.unconfirmed;
}

export function verbCount(delta: DimensionDelta, verb: ChangeVerb): number {
	return delta[verb] ?? 0;
}

File: shared/models/compare.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from shared.definitions.compare import Comparability


class RunSide(BaseModel):
    """One of the two runs."""

    scan_id: uuid.UUID
    engine_name: str
    context_name: str | None = None
    intensity: str = ""
    scope: str = ""
    status: str
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_seconds: float | None = None
    stages_ran: int = 0
    stages_planned: int = 0
    counts: dict[str, int] = Field(default_factory=dict)


class StageDiff(BaseModel):
    name: str
    title: str
    baseline: str | None = None
    current: str | None = None


class SettingDiff(BaseModel):
    stage: str
    title: str
    field: str
    label: str
    before: str | None = None
    after: str | None = None


class RunDifference(BaseModel):
    """One way the two runs were set up differently."""

    key: str
    label: str
    baseline: str | None = None
    current: str | None = None
    material: bool = True


class CoverageLine(BaseModel):
    """One measured account of a run."""

    label: str
    baseline: str | None = None
    current: str | None = None


class DimensionVerdict(BaseModel):
    dimension: str
    comparability: str = Comparability.LIKE_FOR_LIKE.value
    covered_baseline: bool = True
    covered_current: bool = True
    compared: bool = True
    confirmed: bool = True
    note: str = ""
    stages: list[str] = Field(default_factory=list)
    settings: list[SettingDiff] = Field(default_factory=list)
    coverage: list[CoverageLine] = Field(default_factory=list)


class DimensionDelta(BaseModel):
    dimension: str
    label: str
    noun: str
    noun_plural: str
    verdict: DimensionVerdict
    total_baseline: int = 0
    total_current: int = 0
    appeared: int = 0
    changed: int = 0
    disappeared: int = 0
    unconfirmed: int = 0
    unchanged: int = 0
    intel_moved: int = 0

    @property
    def listed(self) -> int:
        return self.appeared + self.changed + self.disappeared + self.unconfirmed


class ChangeField(BaseModel):
    field: str
    label: str
    before: str | None = None
    after: str | None = None
    tone: str = "neutral"


class ScreenshotPair(BaseModel):
    baseline: str | None = None
    current: str | None = None


class ChangeRow(BaseModel):
    key: str
    dimension: str
    verb: str
    signal: str
    rank: int = 99
    title: str
    subtitle: str = ""
    scan_id: uuid.UUID
    fields: list[ChangeField] = Field(default_factory=list)
    severity: str | None = None
    status: int | None = None
    port: int | None = None
    is_kev: bool = False
    sensitive: bool = False
    screenshots: ScreenshotPair | None = None


class ChangeRows(BaseModel):
    dimension: str
    items: list[ChangeRow] = Field(default_factory=list)
    total: int = 0
    page: int = 1
    size: int = 50


class ScanComparison(BaseModel):
    target_id: uuid.UUID
    target_value: str
    target_type: str
    baseline: RunSide
    current: RunSide
    dimensions: list[DimensionDelta] = Field(default_factory=list)
    stage_diff: list[StageDiff] = Field(default_factory=list)
    setting_diff: list[SettingDiff] = Field(default_factory=list)
    settings_identical: int = 0
    run_diff: list[RunDifference] = Field(default_factory=list)
    runs_between: int = 0
    comparability: str = Comparability.LIKE_FOR_LIKE.value
    headline: str = ""
    summary: str = ""
    changes_total: int = 0
    live: bool = False
    suggestion: "ComparableRun | None" = None
    generated_at: datetime


class ComparableRun(BaseModel):
    """A run this one can be put beside."""

    scan_id: uuid.UUID
    engine_name: str
    status: str
    scope: str
    started_at: datetime | None = None
    duration_seconds: float | None = None
    counts: dict[str, int] = Field(default_factory=dict)
    dimensions: list[str] = Field(default_factory=list)
    comparable: bool = True
    reason: str = ""


ScanComparison.model_rebuild()
